src/zarr/core/group.py

```diff
@@ -2765,6 +2836,327 @@ def array(
+                raise ValueError(f"Invalid zarr_format: {zarr_format}")  # pragma: no cover
+
+            coros = (func(store_path=store_path / key) for key in nodes_parsed)
+            extant_node_query = dict(
+                zip(
+                    nodes_parsed.keys(),
+                    await asyncio.gather(*coros, return_exceptions=True),
+                    strict=False,
+                )
+            )
+
+            for key, value in extant_node_query.items():
+                if isinstance(value, BaseException):
+                    if isinstance(value, KeyError):
+                        # ignore KeyErrors, because they represent nodes we can safely create
+                        pass
+                    else:
+                        # Any other exception is a real error
+                        raise value
+                else:
+                    # this is a node that already exists, but a node with this name was specified in
+                    # nodes_parsed.
+                    # Two cases produce exceptions:
+                    # 1. The node is a group, and a node with this name was explicitly defined in
+                    # nodes
+                    # 2. The node is an array.
+                    # The third case is when this extant node is a group, but its name was not
+                    # explicitly defined in nodes. This means it was added as an implicit group by
+                    # _parse_hierarchy_dict, and we can remove the reference to this node from
+                    # nodes_parsed. We don't need to create this node.
+
+                    if isinstance(value, GroupMetadata):
+                        if key not in implicit_group_names:
+                            raise ContainsGroupError(store_path.store, key)
+                        else:
+                            # as there is already a group with this name, we should not create a new one
+                            redundant_implicit_groups.append(key)
+                    elif isinstance(value, ArrayV2Metadata | ArrayV3Metadata):
+                        raise ContainsArrayError(store_path.store, key)
+
+            nodes_parsed = {
+                k: v for k, v in nodes_parsed.items() if k not in redundant_implicit_groups
+            }
+
+    async for node in create_nodes(store_path=store_path, nodes=nodes_parsed, semaphore=semaphore):
+        yield node
+
+
+async def create_nodes(
+    *,
+    store_path: StorePath,
+    nodes: dict[str, GroupMetadata | ArrayV2Metadata | ArrayV3Metadata],
+    semaphore: asyncio.Semaphore | None = None,
+) -> AsyncIterator[AsyncGroup | AsyncArray[ArrayV2Metadata] | AsyncArray[ArrayV3Metadata]]:
+    """
+    Create a collection of zarr arrays and groups concurrently and atomically. To ensure atomicity,
+    no attempt is made to ensure that intermediate groups are created.
+    """
+    ctx: asyncio.Semaphore | contextlib.nullcontext[None]
+
+    if semaphore is None:
```

Comment:
why do we need `None` as an option here?

Reply:
`create_nodes` is low-level, kind of dangerous, and the expectation is that the user of this function knows what they are doing. So it doesn't default with any concurrency limit. It also doesn't check of nodes already exist, of the user wants to nest arrays inside arrays. A higher level function (`create_hierarchy`) does all that checking, and that's where the concurrency limit defaults to the value in the config.

Reply:
so why not require the semaphore if it's an advanced user?

Reply:
the semaphore is only necessary if you want some concurrency limiting mechanism, but I think that's a special case. A default of `None` means that users of the function don't need to run `from asyncio import Semaphore` before creating some nodes. E.g., there are a lot of places in the tests where I would want to use `create_nodes`, and in basically none of those places would a semaphore be necessary.